Create a travis script to build with latest LLVM

jru opened the pull request:
Hi,

After some fighting I got the Travis integration working. Phew!

You can check the result of the latest run (as configured on my fork) here: 
https://travis-ci.org/jru/include-what-you-use/builds/285073561

By default it will build PRs and master immediately after pushing. You just need to configure you account and enable the iwyu repo by going to https://travis-ci.org/

I figured it would be most useful if we keep building against the latest LLVM snapshot (as published in apt.llvm.org) so that we can react proactively should something break. What do you think?

Cheers.

kimgr commented:
Thank you! This looks cool.

A few questions:

- I see this builds on some Ubuntu -- do we know which version?
- LLVM/Clang are installed with a version spec. Does that mean they'll stick to v6.0 even after trunk has moved to 7.0? Can we omit the version, or does that cause other problems?
- Rather than pulling CMake off the web, is there a repo with recent releases we could point to under `sources` instead?
- How hard is it to add support for another platform (Windows)? Would that go in the same travis.yml? I'm not suggesting we do it now, but I'd like to know that it's possible and how to do it

I know preciously little about Travis, so these may be stupid questions :)

janisozaur commented:
> * I see this builds on some Ubuntu -- do we know which version?

14.04.5 "Trusty". It is the default on Travis (https://docs.travis-ci.com/user/reference/overview/#Virtualization-environments). You can also see this under `Build system information`: https://travis-ci.org/jru/include-what-you-use/builds/285073561#L18 and guess from the installed repo name: `llvm-toolchain-trusty`

> * LLVM/Clang are installed with a version spec. Does that mean they'll stick to v6.0 even after trunk has moved to 7.0? Can we omit the version, or does that cause other problems?

The change explicitly requests llvm-6.0. It is required to specify the version, if you want anything else than the default. The way packages are done imposes the version information in them, usually `llvm` would be the transitive package pulling the default, e.g. `llvm-3.9`. Do note that you can have different ways to install packages based on which kind of virtualisation you request from Travis. In this case it is container, which means you can only install [whitelisted packages](https://github.com/travis-ci/apt-package-whitelist) from [whitelisted repos](https://github.com/travis-ci/apt-source-whitelist).

> * Rather than pulling CMake off the web, is there a repo with recent releases we could point to under sources instead?

See previous point and note no `cmake` other than default in here: https://github.com/travis-ci/apt-package-whitelist/blob/master/ubuntu-trusty

> * How hard is it to add support for another platform (Windows)? Would that go in the same travis.yml? I'm not suggesting we do it now, but I'd like to know that it's possible and how to do it

Travis only [offers Linux and macOS workers](https://docs.travis-ci.com/user/multi-os/), with latter being run on much smaller farm due to licensing and cost reasons. There is, however, another service similar to Travis (also free for open-source): https://www.appveyor.com/ which offers Windows VMs for CI purposes. I have no idea how hard it is to get clang/llvm on Windows, but they offer some pre-installed versions too, currently 4.0.0: https://www.appveyor.com/docs/build-environment/#llvm

jru commented:
Thanks for the info @janisozaur. 

I was pretty much settled on having to provide the version but after experimenting a bit I managed to get the latest LLVM picked by the script using some bash magic. The most recent LLVM seems to be installed by default if the version is left out but it is not set as the default. It feels hacky but it works :)

Regarding CMake, I could not find any apt source that supplies 3.4.3 or greater, so I decided to go the easy way and download it directly.

In the meantime I will try AppVeyor. I don't think it brings so much now given that the test suite does not run clean, but at least we could build it.

kimgr commented on .travis.yml:
```diff
@@ -0,0 +1,33 @@
+
+script:
+# Build IWYU
+ - mkdir build && cd build && cmake -GNinja -DIWYU_LLVM_ROOT_PATH=$ROOT_PATH -DCMAKE_C_COMPILER=$BIN_PATH/clang -DCMAKE_CXX_COMPILER=$BIN_PATH/clang++ -DCMAKE_INSTALL_PREFIX=./ ../
```
Could you split this line into the three steps (mkdir, cd, cmake) instead? Or does the one-liner add something?

jru replied:
Each line in the script block will be executed even if the previous failed. So, it adds nothing really for *mkdir* and *cd* since they can't fail, but for *cmake* and *ninja* I think it makes sense. I'll change it as you suggested

kimgr commented on .travis.yml:
```diff
@@ -0,0 +1,33 @@
+ - sudo sh cmake.sh --skip-license --exclude-subdir --prefix=/usr/local
+
+ # Extract the version number from the most-recently installed LLVM
+ - VERSION=`ls -t /usr/lib/ | grep '^llvm\-' | head -n 1 | sed 's/llvm\-\(.\+\)/\1/'`
```
The escaping of `\+` looks wrong here, but apparently it works...? I would guess `\(.+\)` would be a functional capturing group. Likewise, the escaping of `-` is probably not necessary in the sed expression (grep requires it, IIRC).

jru replied:
Yeah, I think can simplify it by using the extended regex mode (-E). I never remember which mode escapes what.

kimgr commented on .travis.yml:
```diff
@@ -0,0 +1,33 @@
+ - ninja install
+
+# Test IWYU
+ - cd .. && python run_iwyu_tests.py -- $TRAVIS_BUILD_DIR/build/bin/include-what-you-use
```
Since you must be in `$TRAVIS_BUILD_DIR` at this point, I wonder if it's necessary to spell it out. Maybe just `build/bin/include-what-you-use`?

kimgr replied:
Also, could this be split onto two lines?

kimgr commented on README.md:
```diff
@@ -1,3 +1,5 @@
+[![Build Status](https://travis-ci.org/jru/include-what-you-use.svg?branch=master)](https://travis-ci.org/jru/include-what-you-use)
```
Reminder for myself: this needs to be updated before we merge, to point to an "official" IWYU build.

jru replied:
I can do that too!

jru commented:
Hi,

We could definitely find a way to skip the failing tests directly in the build script, but I think it would be much easier to extend the runner to ignore certain tests passed by argument.

Regarding PRs, yes, that script will be run for all of them and there will be some notifications sent informing of failure or success. That should make reviewing small changes much easier as you won't necessarily need to build & test locally.

I will address your comments soon! Thanks.

jru commented:
Done. 

Btw, the Travis build seems to be failing now due to the llvm package being unavailable (404). I hope is a transient issue..

janisozaur commented:
> How does this work, admin-wise -- does travis-ci.org pick up automatically or do we need to register an account, or something?

You need to _register_ with Travis. Note that it will use your GitHub login information, so there's no need to create another login. I think it can be anyone who is a member of https://github.com/include-what-you-use organization. Then, go hover over your login, go to "Accounts" and switch to "include-what-you-use" and enable this repository.

It should automatically pick up any new commits to the repository and new PRs (do check in options that they are enabled; as this one is already created, it won't get picked up unless re-PR'd).

Do note the above statement means that PRs coming from branches inside this repository will get two distinct runs – one "push" and one "PR".

jru commented on .travis.yml:
```diff
@@ -0,0 +1,37 @@
+# Test IWYU
+ - cd ..
+ - python run_iwyu_tests.py -- build/bin/include-what-you-use
+
```
Sure thing!

jru commented on .travis.yml:
```diff
@@ -0,0 +1,37 @@
+ - mkdir build
+ - cd build
+ - cmake -GNinja -DIWYU_LLVM_ROOT_PATH=$ROOT_PATH -DCMAKE_C_COMPILER=$BIN_PATH/clang -DCMAKE_CXX_COMPILER=$BIN_PATH/clang++ -DCMAKE_INSTALL_PREFIX=./ ../
+ - ninja install
```
It searches both, in the system dirs and in the installation directory. Not sure which ones, but I do rememeber seeing it fail if I didn't install it.

kimgr replied:
OK, that explains it, thanks!

kimgr commented:
Thanks, I'll register now.

> Do note the above statement means that PRs coming from branches
> inside this repository will get two distinct runs – one "push" and one "PR".

We typically don't create PRs from this repo, iwyu maintainers have forks and PR from there, so this should not be an issue. But thanks for bringing it up, I could see myself getting confused!

kimgr commented:
There, I have registered `include-what-you-use/include-what-you-use` now, to only build if `.travis.yml` is available. So it *should* pick up when this is merged.

jru commented:
Just updated the PR with the corrections and Jenkins picked it up already. First green build :)

I added fix_includes_test and squashed the corrections. Should I also mention in the commit that it fixes #293?

kimgr commented:
Very nice! Thank you for working on this.
